coderd/httpmw/workspaceproxy.go

```diff
@@ -0,0 +1,158 @@
+package httpmw
+
+import (
+	"context"
+	"crypto/sha256"
+	"crypto/subtle"
+	"database/sql"
+	"net/http"
+	"strings"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/database/dbauthz"
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/codersdk"
+)
+
+const (
+	// ExternalProxyAuthTokenHeader is the auth header used for requests from
+	// external workspace proxies.
+	//
+	// The format of an external proxy token is:
+	//     <proxy id>:<proxy secret>
+	//
```

Comment:
I think it's better to use a different format because they're incompatible TBH and it'll make it easy to tell if people are trying to use proxy tokens as API keys etc.

Reply:
I don't disagree with the ease of telling a difference, it just seems weird to maintain multiple types of secret keys when we already have a standard in the product.

If we need to bump the security of our keys, we'll now have multiple places for that to be done, even though they share the same purpose.

And under this format, it's not explicit when it's a proxy key... we just know it is because of the `:`, right?

Reply:
We already have multiple places where keys would need to be bumped because agent tokens are not handled with the same code as API keys so I don't think this is a concern. The existing format wasn't used because it doesn't make sense to have two IDs for a workspace proxy. We're already using a UUID for workspace proxies, so IMO it makes sense to use that as the key ID as well. The proxy keys are not stored in a different table since each proxy only has one key.

It's a lot longer than a regular key and has a `:`, it's a UUID instead of 10 char ID so it would be pretty easy to tell apart.